ddl/delete_range.go

```diff
@@ -0,0 +1,334 @@
+func (dr *delRange) start() {
+	if !dr.storeSupport {
+		go dr.startEmulator()
+	}
+}
+
+// clear implements delRangeManager interface.
+func (dr *delRange) clear() {
+	log.Infof("[ddl] closing delRange session pool")
+	dr.ctxPool.Close()
+}
+
+func (dr *delRange) startEmulator() {
+	log.Infof("[ddl] start delRange emulator")
+	dr.d.wait.Add(1)
+	defer dr.d.wait.Done()
+	for {
+		select {
+		case <-dr.emulatorCh:
+		case <-dr.d.quitCh:
+			return
+		}
+		dr.doDelRangeWork()
+	}
+}
+
+func (dr *delRange) doDelRangeWork() error {
+	resource, err := dr.ctxPool.Get()
+	if err != nil {
+		log.Errorf("[ddl] delRange emulator get session fail: %s", err)
+		return errors.Trace(err)
+	}
+	defer dr.ctxPool.Put(resource)
+	ctx := resource.(context.Context)
+	ctx.GetSessionVars().SetStatusFlag(mysql.ServerStatusAutocommit, true)
+
+	ranges, err := LoadDeleteRanges(ctx, math.MaxInt64)
+	if err != nil {
+		log.Errorf("[dd] delRange emulator load tasks fail: %s", err)
+		return errors.Trace(err)
+	}
+
+	for _, r := range ranges {
+		if err := dr.doTask(r); err != nil {
+			log.Errorf("[ddl] delRange emulator do task fail: %s", err)
+			return errors.Trace(err)
+		}
+		if err := CompleteDeleteRange(ctx, r); err != nil {
+			log.Errorf("[ddl] delRange emulator complete task fail: %s", err)
+			return errors.Trace(err)
+		}
+		log.Infof("[ddl] delRange emulator complete task: (%d, %d)", r.jobID, r.elementID)
+	}
+	return nil
+}
+
+func (dr *delRange) doTask(r DelRangeTask) error {
+	for {
+		finish := true
+		dr.keys = dr.keys[:0]
+		err := kv.RunInNewTxn(dr.d.store, false, func(txn kv.Transaction) error {
+			iter, err := txn.Seek(r.startKey)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			defer iter.Close()
+
+			for i := 0; i < delBatchSize; i++ {
+				if iter.Valid() {
+					finish = bytes.Compare(iter.Key(), r.endKey) >= 0
+					if !finish {
+						dr.keys = append(dr.keys, iter.Key().Clone())
+						if err := iter.Next(); err != nil {
+							return errors.Trace(err)
+						}
+						continue
+					}
+				}
+				break
+			}
+			for _, key := range dr.keys {
+				err := txn.Delete(key)
+				if err != nil && !terror.ErrorEqual(err, kv.ErrNotExist) {
+					return errors.Trace(err)
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if finish {
+			break
+		}
+	}
+	return nil
+}
+
+// insertBgJobIntoDeleteRangeTable parses the job into delete-range arguments,
+// and inserts a new record into gc_delete_range table. The primary key is
+// job ID, so we ignore key conflict error.
+func insertBgJobIntoDeleteRangeTable(ctx context.Context, job *model.Job) error {
+	s := ctx.(sqlexec.SQLExecutor)
+	switch job.Type {
+	case model.ActionDropSchema:
+		var tableIDs []int64
+		if err := job.DecodeArgs(&tableIDs); err != nil {
+			return errors.Trace(err)
+		}
+		for _, tableID := range tableIDs {
+			startKey := tablecodec.EncodeTablePrefix(tableID)
+			endKey := tablecodec.EncodeTablePrefix(tableID + 1)
```

Comment:
According to document of `PrefixNext`, I'm not sure it can be used here. So I will test it later.